llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -9882,12 +9895,22 @@ RISCVTargetLowering::lowerFixedLengthVectorStoreToRVV(SDValue Op,
 
   MVT ContainerVT = getContainerForFixedLengthVector(VT);
 
-  SDValue VL = getVLOp(VT.getVectorNumElements(), ContainerVT, DL, DAG,
-                       Subtarget);
-
   SDValue NewValue =
       convertToScalableVector(ContainerVT, StoreVal, DAG, Subtarget);
 
+
+  // If we know the exact VLEN and our fixed length vector completely fills
+  // the container, use a whole register store instead.
+  const auto [MinVLMAX, MaxVLMAX] =
+      RISCVTargetLowering::computeVLMAXBounds(ContainerVT, Subtarget);
+  if (MinVLMAX == MaxVLMAX && MinVLMAX == VT.getVectorNumElements() &&
+      getLMUL1VT(ContainerVT).bitsLE(ContainerVT))
+    return DAG.getStore(Store->getChain(), DL, NewValue, Store->getBasePtr(),
+                        Store->getMemOperand());
```

Comment:
I'm seeing a miscompile on SPEC CPU 2017 502.gcc_r when compiled with `-march=rv64gv -mrvv-vector-bits=zvl -O3` that was bisected back to here e8a15eca92f1a10b3af4f4e52f54d9d2d7612bf5. It seems to be caused by this store part and not the load, *only* with SEW=64 and LMUL=1. 

I don't have any reduced test case to share at the moment, and looking at the code and the test diffs nothing seems to stick out to me. Will keep investigating

Reply:
The ultimate cause of the miscompile was due to a dodgy combine on ISD::STORE which incorrectly removed a load/store pair, assuming they didn't alias. There was nothing wrong in this PR, it just happened to trigger it by lowering to an ISD::STORE node instead of a llvm.riscv.vse intrinsic.

Coincidentally on the same day I began investigating this, https://github.com/llvm/llvm-project/pull/83017 was landed which fixes it.

I've added a test case here which illustrates it in more detail: 63725ab1196ac50509ad382fc12c56f6d8b5d874